egs/librispeech/asr/simple_v1/mmi_bigram_embeddings_decode.py

```diff
@@ -0,0 +1,268 @@
+#!/usr/bin/env python3
+
+# Copyright (c)  2020  Xiaomi Corporation (authors: Daniel Povey, Haowen Qiu)
+# Apache 2.0
+
+import k2
+import logging
+import numpy as np
+import os
+import torch
+from k2 import Fsa, SymbolTable
+from kaldialign import edit_distance
+from pathlib import Path
+from typing import List
+from typing import Union
+
+from lhotse import CutSet
+from lhotse.dataset import K2SpeechRecognitionDataset, SingleCutSampler
+from snowfall.common import find_first_disambig_symbol
+from snowfall.common import get_texts
+from snowfall.common import load_checkpoint
+from snowfall.common import setup_logger
+from snowfall.decoding.graph import compile_LG
+from snowfall.models import AcousticModel
+from snowfall.models.tdnn_lstm import TdnnLstm1b
+from snowfall.training.ctc_graph import build_ctc_topo
+from snowfall.training.mmi_graph import create_bigram_phone_lm
+from snowfall.training.mmi_graph import get_phone_symbols
+
+
+def decode(dataloader: torch.utils.data.DataLoader, model: AcousticModel,
+           device: Union[str, torch.device], LG: Fsa, symbols: SymbolTable):
+    tot_num_cuts = len(dataloader.dataset.cuts)
+    num_cuts = 0
+    results = []  # a list of pair (ref_words, hyp_words)
+    for batch_idx, batch in enumerate(dataloader):
+        feature = batch['features']
+        supervisions = batch['supervisions']
+        supervision_segments = torch.stack(
+            (supervisions['sequence_idx'],
+             torch.floor_divide(supervisions['start_frame'],
+                                model.subsampling_factor),
+             torch.floor_divide(supervisions['num_frames'],
+                                model.subsampling_factor)), 1).to(torch.int32)
+        indices = torch.argsort(supervision_segments[:, 2], descending=True)
+        supervision_segments = supervision_segments[indices]
+        texts = supervisions['text']
+        assert feature.ndim == 3
+
+        feature = feature.to(device)
+        # at entry, feature is [N, T, C]
+        feature = feature.permute(0, 2, 1)  # now feature is [N, C, T]
+        with torch.no_grad():
+            nnet_output = model(feature)
+        # nnet_output is [N, C, T]
+        nnet_output = nnet_output.permute(0, 2,
+                                          1)  # now nnet_output is [N, T, C]
+
+        #  blank_bias = -3.0
+        #  nnet_output[:, :, 0] += blank_bias
+
+        dense_fsa_vec = k2.DenseFsaVec(nnet_output, supervision_segments)
+        # assert LG.is_cuda()
+        assert LG.device == nnet_output.device, \
+            f"Check failed: LG.device ({LG.device}) == nnet_output.device ({nnet_output.device})"
+        # TODO(haowen): with a small `beam`, we may get empty `target_graph`,
+        # thus `tot_scores` will be `inf`. Definitely we need to handle this later.
+        lattices = k2.intersect_dense_pruned(LG, dense_fsa_vec, 20.0, 7.0, 30,
```

Comment:
Am I right that this is just standard decoding right now, i.e. we're not doing some kind of n-best from a 2nd-pass decoding and then doing voting?

Reply:
Yes, you are right. This script is renamed from `mmi_bigram_decode.py`.

Reply:
> we're not doing some kind of n-best from a 2nd-pass decoding and then doing voting?

@danpovey  Could you explain the meaning of `voting` in this context?